Update release/deployment RFC (#22)
wellcomecollection/docs: rfcs/013-release_deployment_tracking/README.md

## Background

We have multiple projects that are themselves composed of multiple services, which can deploy into multiple environments (e.g. production/staging).

Folk are regularly making changes, building updated services and deploying them into the appropriate environments. Keeping track of what is deployed where, by who and why is essential to working effectively.

In general the build/release/deployment process can be described as follows:

![Simple overview](simple_flow.png)

More specifically including a high level view of infrastructure includes:

- A build environment to build artifacts, e.g. Travis
- An artifact store to store things, e.g. ECR
- A database to keep track of what should be released where e.g. DynamoDB

> - A Deployment environment to run things in, e.g. ECS

Reviewer: I’m not sure this is adding anything over your diagram; if you want to keep it I'd flip it around to put the description first and the document-specific term second:

```suggestion
- A service that creates build artefacts from a given version of the codebase, e.g. creating a Docker image (a *build environment*)
- A store for the created artefacts, e.g. Docker images (an *artefact store*)
- An environment where services can run, e.g. ECS or Kubernetes (a *deployment environment*)
- A database that tracks what version of each application is running
```

![Infrastructure overvier](high_level_infra.png)

### Glossary

> In order that we can talk about the numerous concepts flying about here, we'll define some terms and visualise how they fit together:

Reviewer: You can ditch this sentence; the purpose of a glossary is common knowledge.

- **project:** A high level abstraction, consisting of one or more **service set**'s. Practically this might indicate whole product and should be a single git repository, e.g. the "catalogue" project.

- **service set:** A functional grouping of **services** within a project. e.g. all the services for the catalogue pipeline. You can have multiple per project, for example in the catalogue project, you've got pipeline, api and adapters.

> - **service:** Performs a distinct function within a **project**, practically it may be composed of a few closely related containers. e.g. id_minter, requests API, Front-end content app.

Reviewer: A project is made up of service sets, a service set is made of services, a service is part of a project. It’s a circular definition!

I would flip the order here: service, then service set, then project. Build up the understanding from small to big. Don't refer forwards, only backwards – that way I don't have to jump around to understand.

> - **service:**: an application that performs a single function. For example, the catalogue API is a service that reads documents from Elasticsearch and presents them to a user. [Details of containers, composition, etc. are not necessary here.]
> - **service set:** a collection of services that are deployed together. For example, the catalogue pipeline is a service set of all the applications that populate the works index in Elasticsearch. [Simply giving names as examples isn't enough, because (1) it's prone to change and (2) it makes it harder for people unfamiliar with our projects to follow.]

And having written those two out, I wonder: what's the point of a project? Is that a useful concept?

> - **environment:** Where you deploy your **services** when you want them to run! e.g. staging, production

Reviewer: This definition could equally apply to "ECS or Kubernetes" – it would be good to explain that environments should look broadly similar, but serve different purposes, e.g. prod is for live traffic, staging is for testing.

> - **build environment:** Where you build artifacts to deploy, for example CircleCI, Travis, or your local machine (if tooling allows).

Reviewer: > Where you create an artefact that can be run as a service.

(Is this a useful concept?)

Author: I think we can safely remove it.

> - **build:** The process of creating a **build artifact** for a single **service**

Reviewer: Definition of build should precede build environment.

> - **build artifact:** A deployable _thing_ for a single **service**, practically this is a container image stored in ECR.

Reviewer: Definition of build artefact should precede build.

> - **release hash:** Metadata uniquely identifying a **build artifact**, practically this will be a git ref. 

Reviewer: How does a release hash identify a build artefact? Why do we need it?

```suggestion
- **release hash:** Metadata that allows us to work out what version of the code was used to create a given build artifact. For example, the Git commit hash.
```

"hash" is a very Git-specific term. What happened to "release ID"?

- **release:** Metadata indicating the intention to deploy a particular **build artifact** at a given **release hash**. Generally part of a **release set**.

> - **release set:** A set of **build artifacts** at particular **release hashes** based on a **service set** template that is intended to be released into an **environment** together.

Reviewer: This is better! release hash > release > release set, follows much more smoothly.

## Background

We run a lot of applications. People are regularly making changes and deploying new code. In order that we don't get confused about what code is deployed where we should track what is deployed where, by whom, and for what reason.

The build/release/deployment process can be described as follows:

![Simple overview](simple_flow.png)

A high level view of infrastructure includes:

- A service that creates build artifacts from a given version of the codebase, e.g. creating a Docker image (a *build environment*)
- A store for the created artifacts, e.g. Docker images (an *artifact store*)
- An environment where services can run, e.g. ECS or Kubernetes (a *deployment environment*)
- A database that tracks what version of each application is running

![Infrastructure overvier](high_level_infra.png)

### Glossary

- **project:** The top level, consisting of one or more **service set**'s. This might indicate whole product and should be a single git repository, e.g. the catalogue project.

> - **service:** Performs a distinct function within a **project**. e.g. id_minter, requests API, Front-end content app.

Reviewer: I completely understand what you mean here, but it may not be clear to somebody who's unfamiliar with our stuff already.

How about:

“This could be a single step of a multi-stage pipeline, an API application, or a front-end content app.”

- **service set:** A functional grouping of **services** within a project. You can have multiple per project, for example in the catalogue project, you've got pipeline, api and adapters.

- **build:** The process of creating a **build artifact** for a single **service**

> - **build artifact:** A deployable _thing_ for a single **service**, e.g. a docker image or zip file.

Reviewer: ```suggestion
- **build artifact:** A deployable _thing_ for a single **service**, e.g. a Docker image or Lambda zip file.
```

- **release hash:** Metadata that allows us to work out what version of the code was used to create a given build artifact, e.g. the Git commit hash.

- **release:** Metadata indicating the intention to deploy a particular **build artifact** at a given **release hash**. Generally part of a **release set**.

- **release set:** A set of **build artifacts** at particular **release hashes** based on a **service set** template that is intended to be released into an **environment** together.

- **deployment:** A deployed **service**.

- **environment:** Where you deploy your **release sets** when you want them to run e.g. staging, production.

- **deployment set:** A set of deployed **services** created from a **release set** that has been deployed into an **environment**.

#### How these terms fit together

> ![Terms](terms.png)

Reviewer: nit: “deplyment set”

## What we do now

See the documentation on [version 1](v1/README.md).

### Problems with the current approach

> - It is not clear how to release a single service

Reviewer: ```suggestion
- It is not clear how to release a single service, as opposed to releasing an entire service set
```

- In order to actually deploy something there are multiple steps:
  - Create a release set using the CLI tool
  - Deploy a release set  using the CLI tool
  - Run `terraform apply` to _actually_ update the running services
- Release/Deploy descriptions are not well used / hidden
- Poor visibility of what is actually deployed

#### Moving away from terraform for deployment

We currently use `terraform apply` to deploy services at a particular release hash. The choice to use terraform was driven by a requirement to describe our task definitions in code. 

Separating service deployment from infrastructure changes is desirable as infra/service deployments have differing concerns and pace, i.e. high-value infrequent (infra), vs. low-value frequent (deploying new versions of services). 

> Running terraform in a CI environment like Travis is not desirable as giving an automated environment the power to run infrastructure updates needs careful consideration.

Reviewer: The point of writing RFC is to give careful consideration to these things!

We have discussed reasons we don't want to give Travis credentials:

* risk of credentials being leaked or misused
* we don't usually have permanent credentials
* we've found an alternative

And possibly others I've forgotten.

Also, use more active voice. “We don’t want to give Travis the power to make infrastructure changes because XYZ”, not “it is not desirable”. It's more direct, and we're stating our opinion, not an objective fact.

#### Why this is hard

> An ECS task definition contains configuration for volume mounts, CPU & memory requirements, as well as indicating the container image URI to use when creating tasks.

Reviewer: this might be clearer as a bulleted list

> The container image URI cannot be updated independently from other parameters in a task definition. This makes ignoring a change to the task definition difficult, see this [epic GitHub issue thread](https://github.com/terraform-providers/terraform-provider-aws/issues/632).

Reviewer: “The container image URI cannot be updated independently from other parameters in a task definition” ~ is this a terraform-specific problem? Pretty sure it's possible with the ECS API, the dance with getting the old task definition first aside.

When terraform updates a task definition it has a version of the task definition in code to send to ECS, the ECS Service is then updated by terraform to point at that new task definition and a deployment is started in ECS.

> However if the task definition is updated and differs from that recorded by the terraform state (which updating the image URI _would_ cause) terraform will attempt to return the task definition to a known state, which would be undesirable.

Reviewer: I think I understand these sentences, but they're not so clear. Here's how I parsed it:

> When Terraform updates an ECS task definition, it sends the task definition to ECS and records that in the Terraform state. Terraform then tells the ECS service to use the new task definition, and ECS starts appropriate containers for the task.
>
> If the task definition is updated outside Terraform (including changing the image URI), then Terraform will attempt to revert the task definition to the version in its state. This could cause a service to go backwards.

In order to move away from using `terraform apply` it will be necessary to decouple updating the task definition from deploying updated services.

### Proposed Solution

We intend to address the problems described above my improving on the existing CLI tool.

We will:

> - Provide complete documentation with examples for the updated CLI tool

Reviewer: ```suggestion
- Provide complete documentation with examples for the updated CLI tool, so it is easier to see how to deploy both single services and a complete service set
```

> - Provide "single step" deployment capability in the CLI tool
> - Remove the requirement to run `terraform apply` to update existing services

Reviewer: These two are really the same thing, I think?

```suggestion
- Provide "single step" deployment capability in the CLI tool.
  In particular, we will remove the requirement to run `terraform apply` to update existing services.
```

- Provide quick visibility on the current state of deployments
- Remove or automate "descriptions" required from users of the CLI tool

#### General approach

We'll use consistent image URIs in task definitions and update what those URIs reference instead of updating image URIs, allowing us to keep task definitions static when updating which container images they should use. 

As the relationship between which container image to use in which service is no longer described as part of the infrastructure we can avoid terraform.

Docker container image repositories allow us to do this through the use of tags. A particular docker image can have multiple tags, and we can use this to provide "environment based tags", e.g. prod, stage. We update what tags are attached to which image to indicate what should be deployed where.

> In order that tag changes are "noticed" by our deployed services we will need to force redeployment of the correct ECS services after updating tags. This will cause the tasks created by the new deployment to read their container images from their new tag.

Reviewer: This section might benefit from some examples:

> Previously the image URI in our task definition was the Git commit hash of the code used to build the Docker image. Now we will use a consistent (or maybe “stable”?) image URI, which will point to different images over time.
>
> Before: `our_great_app:b17af8e`
> After: `our_great_app:latest`
>
> From Terraform's point of view, the image URI never changes, so it won't try to change anything.

![General approach](general_approach.png)

#### Register images

When container images are built we should tag them with:
- **release hash** as we do now, to keep track of the relationship between release hash and container image id. This tag will always be associated with the same container image.
- **latest** so we have a way of knowing the most recently created image for each service which is useful when planning deployments. This tag should always be on the most recently built image for a service.

If you were to use the docker CLI tool, this would look like:

```sh
# Create tags!
docker tag image_i_just_built ecr_repo/service_name:hash_1

> docker ecr_repo/service_name:hash_1 ecr_repo/service_name:latest

Reviewer: ```suggestion
docker tag image_i_just_built ecr_repo/service_name:latest
```

# Push tags!
docker push ecr_repo/service_name:hash_1
docker push ecr_repo/service_name:latest
```

The release CLI tool should automate this process so that we encode this logic in one place. 

#### Deploying

Deploying is now a process of:

>  - Ascertaining which services you wish to deploy

Reviewer: /s/ascertaining/deciding?

## Background

We should track what code is deployed where, by whom and for what reason. This will give us a clear picture of the state of our deployments, which is useful for tracking the progress of bugfixes and new features.

The build/release/deployment process can be described as follows:

> ![Simple overview](simple_flow.png)

Reviewer 2: I know it’s pretty simple, but I really like the `build pushes to release tool` and `deploy reads from release tool`.
